consensus/propagation/reactor.go

```diff
@@ -465,14 +475,61 @@ func (blockProp *Reactor) handleWants(peer p2p.ID, wants *proptypes.WantParts) {
 	_, parts, _, has := blockProp.GetProposal(height, round)
 	// the peer must always send the proposal before sending parts, if they did
 	//  not, this node must disconnect from them.
-	if !has {
+	// TODO fix this hack where we check for round -2
+	if round != -2 && !has {
 		blockProp.Logger.Error("received part state request for unknown proposal", "peer", peer, "height", height, "round", round)
 		// d.pswitch.StopPeerForError(p.peer, fmt.Errorf("received part state for unknown proposal"))
 		return
 	}
 
 	// send the peer the partset header if they don't have the proposal.
 	// TODO get rid of this catchup case
+	if round == -2 {
```

Comment:
general code-smell of having a very large chunk of code behind an if statement.

it might be more readable if we check for and handle the simple case first. adding other specific message types would also handle this.

Reply:
yes 100%, I still didn't push the new changes. This was just a fast draft to see if things are working.